File: Codility/10_prime_and_composite_numbers/peaks.rb
# https://codility.com/programmers/task/peaks/

def is_valid_solution(k, peaks, n)
  peak_index = 0
  (0..n - 1).step(k) do |i|
    return false if peak_index >= peaks.length
    return false if peaks[peak_index] >= i + k
        
    while peak_index < peaks.length and peaks[peak_index] < i + k
      peak_index += 1
    end
  end
  
  true
end

def solution(a)
  n = a.length
  peaks = a[1..n - 1].each_with_index.map { |n, i| i if a[i] > a[i - 1] and a[i] > a[i + 1] }.compact
  
  # puts peaks
  
  for k in 2..n/2
    return n/k if n % k == 0 and is_valid_solution(k, peaks, n)
  end
  
  return 1 if is_valid_solution(n, peaks, n)
  
  0
end

File: Codility/10_prime_and_composite_numbers/peaks.cpp
// https://codility.com/programmers/task/peaks/

bool isValidDivision(const vector<size_t>& peaks, size_t k, size_t N)
{
    auto it = peaks.cbegin();
    for (size_t i = 0; i < N; i+= k)
    {
        if (it == peaks.cend())
            return false;
        
        if (*it >= i + k)
            return false;
            
        while (it != peaks.cend() && *it < i + k)
            ++it;
    }
    
    return true;
}

int solution(vector<int> &A) {
    const size_t N{A.size()};
    vector<size_t> peaks;
    for (size_t i = 1; i < N - 1; ++i)
    {
        if (A[i - 1] < A[i] && A[i] > A[i + 1])
            peaks.push_back(i);
    }
    
    for (size_t k = 2; k <= N / 2; ++k)
    {
        if (N % k)
            continue;
            
        if (isValidDivision(peaks, k, N))
            return N / k;
    }

    if (isValidDivision(peaks, N, N))
        return 1;
    
    return 0;    
}